src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -223,7 +223,7 @@ private StringBuilder GetStringBuilder()
         /// <param name="pathData">The <see cref="VirtualPathData"/>.</param>
         /// <param name="fragment">The URL fragment.</param>
         /// <returns>The generated URL.</returns>
-        protected virtual string GenerateUrl(string protocol, string host, VirtualPathData pathData, string fragment)
+        internal virtual string GenerateUrl(string protocol, string host, VirtualPathData pathData, string fragment)
```

Comment:
This isn't right. This was intentionally protected so developers could override it.

If you need to call this method directly in tests, create a subclass and a public `new` method that calls base